### ISSUE TITLE
idea: make +vendor the default if nothing is specified for fetch.

### ISSUE THREAD

**Issue by pkieltyka** [kardianos/govendor#132]
hello, I thought `govendor sync` would sync all packages with their remote origins? but, I'm finding this isn't happening. what is the command to do:

1. return status list of all outdated packages to their external origin
2. fetch/vendor each outdated package with their external origin

**Comment by kardianos** [comment]
fetch +v
Should work.

**Comment by pkieltyka** [comment]
Great to know :) please update README with a series of common commands

**Comment by kardianos** [comment]
Have you seen the FAQ?

**Comment by pkieltyka** [comment]
Ah, I missed the faq. Just tried a few commands, then checked README, after that opened this issue :) 

going to close the issue, but maybe thats some data/insight for a dev's thought process

**Comment by pkieltyka** [comment]
btw, have you thought to make +v the default flag if one isn't set..? that is, it will act on the vendor'ed packages..? which is what people are thinking is happening when using "govendor", with "vendor" in the name :)

ie.: `$ govendor fetch` would default to `$ govendor fetch +v` .. etc.

**Comment by kardianos** [comment]
I don't want to do that now, but reasonable idea. Reopening issue to record.

**Comment by pkieltyka** [comment]
@kardianos you know my love for govendor, but I think some sensible defaults for the CLI could make it more useful for its users. I personally don't think a dep tool should require documentation, it should need 3 commands, and boom, away you go. just a quick read of govendor -h should tell the person everything the need. Then, if they want to run more advanced operations, they can peal into the tool further for advanced operations.

Perhaps https://github.com/pressly/gov might have some ideas for you, it's a spec I started working on a few months ago before I discovered govendor and decided to drop my project.

I believe, given that `govendor` is written for the vendoring Go deps in "vendor/", then it should default for that.. ie. for one take, here is what naturally comes to mind:

`govendor init`
* check, works great :)

`govendor get [<pkgA>[:tag] ..<pkgB>[:tag]]`
* test each vendor/ dep against the remote source and upgrade it to the latest
* optional arguments are to sync a specific page by its import path. ie. `govendor sync github.com/pressly/chi github.com/goware/jwtauth` would only sync chi and jwtauth packages, if for some reason they dont exist, lets add them and note they've been newly added
* package import paths can also accept an optional SCV tag which can be a commit id, tag or branch

`govendor prune [<pkgA>[:tag] [..<pkgB>[:tag]]
* inform the user about unused packages and remove them from vendor/
* its a way to cleanup the vendor/ directory or remove a specific package

`govendor list [--outdated]`
* print the dep tree in vendor/
* if --outdated is passed, it will test against remote sources and mark deps that can be updated

.. personally, I think these are the only commands you need. I think giving people fewer options makes it more obvious how to use the tool.

One more idea, (optional if you want), is to accept --gopath for the `get` command so that it uses the $GOPATH as the source to grab dependencies from, or update, etc.

**Comment by pkieltyka** [comment]
btw, as I read through the commands, I can see one challenge with `govendor get` which would be when no args are passed then, it could get/fetch all packages in vendor/vendor.json, but the problem here is, if your vendor.json is empty it will have no effect, which perhaps is okay and it should print to say, "either give me a package or run govendor sync <import-path-to-your-main-package>" 

so, perhaps there should be a `govendor sync <import-path-to-your-main-package>` command that is used to sync the initial list, or vendorly newly packages used by a program that previously weren't vendored. I've considered to list <import-path-to-main> because the dep tree must be calculated from some point, and its a popular convention to include a `cmd/myapp/main.go` for a program, and so the tree would be calculated from that point. Ie. `govendor sync github.com/pressly/api/cmd/api-server`

**Comment by kardianos** [comment]
@pkieltyka I'm not against these ideas in general. However it isn't the tool that's been released.

However, govendor isn't just a CLI tool. It is a collection of packages that can be used to work with the vendor folder. The CLI tool happens to use these packages. It would probably only take a few hours to create a new command, based on govendor/run and govendor/context that will do all these things.

So in short, I'd welcome a different take on the CLI, but for now, I think I'd go in the direction of a new CLI tool for now, at least until I've played with it.

**Comment by pkieltyka** [comment]
interesting.. then perhaps I could build "gov" and use govendor... but, I just feel the real utility/credit/focus is deserved by govendor. Or.. how about govendor provide a command like "gov" via github.com/kardianos/govendor/cmd/gov :)

I like how you explain a bigger vision for govendor, and encourage the use of its subpackages to write other tools to make it easier to manage dependencies (perhaps include some of those thoughts in the README too), but, as well, consider how will govendor be used 99% of the time? its probably too early to tell, but, a big part will be as a cli for vendoring pkgs for a project -- in which case, design for the most common use case, the best system design is specific. So, the thought could be to design the cli to be useful for 99% of usecases, but, still have the advanced operations accessible through the CLI in a similar or scoped under some key like `dep` ... as in `govendor dep list` , `govendor dep add +external`.

too many options :)

**Comment by kardianos** [comment]
I'd need to see how the project progressed, but it might be an option to do just that "govendor/cmd/gov".

In my experience there is a common "getting started" experience. But after that having a robust tool is much more important, and the benefit of "govendor list" and "govendor list -v filter" is supper important.

Let me know what you decide!

**Comment by pkieltyka** [comment]
cool :) lets see how everything unfolds. I hear what you mean about the "getting started" hump and learning a more powerful tool can be more valuable than hiding features, thats a valid perspective. But consider that most people are lazy and working too much, so the more intuitive and comfortable the "developer experience" for adopting a technology, the lower the friction, and thus the greater the adoption.
